dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +272,371 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+        for (auto task : tasks)
+        {
+            if (task.Value().Name() == backgroundTaskFullName)
+            {
+                return task.Value();
             }
+        }
+        return nullptr;
+    }
 
-            BackgroundTaskBuilder builder{ nullptr };
+    void PushNotificationManager::Register()
+    {
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_OPERATION_IN_PROGRESS), m_registering, "Registration is in progress!");
+            m_registering = true;
+        }
 
-            if (WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::PushTrigger) && PushNotificationHelpers::IsPackagedAppScenario())
+        try
+        {
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
-
                 {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_pushTriggerRegistration);
+                    auto lock{ m_lock.lock_shared() };
+                    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED), m_comActivatorRegistration || m_pushTriggerRegistration);
                 }
 
-                winrt::hstring taskClsidStr{ winrt::to_hstring(taskClsid) };
-                winrt::hstring backgroundTaskFullName{ backgroundTaskName + taskClsidStr };
-
-                auto tasks{ BackgroundTaskRegistration::AllTasks() };
-                bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
-                    [&](auto&& task)
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
+                {
                     {
-                        auto name{ task.Value().Name() };
+                        auto lock { m_lock.lock_exclusive() };
+                        m_comActivatorRegistration.reset();
 
-                        if (std::wstring_view(name).substr(0, backgroundTaskName.size()) != backgroundTaskName)
+                        // Clean the task registration only if it was created during this call
+                        if (m_pushTriggerRegistration)
                         {
-                            return false;
+                            m_pushTriggerRegistration.Unregister(true);
+                            m_pushTriggerRegistration = nullptr;
                         }
 
-                        if (name == backgroundTaskFullName)
-                        {
-                            s_pushTriggerRegistration = task.Value();
-                            return true;
-                        }
+                        m_registering = false;
+                    }
 
-                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
-                    }) };
+                    UnregisterSinkHelper();
```

Comment:
This will throw if no sinks are registered. Better off having a seperate scope for this right before you call RegisterSinkHelper.

Reply:
@pmpurifoy  or you could also use the m_singletonForegroundRegistration flags?

Reply:
I wouldnt recommend the same flag to override. If you put the new scope next to the actual RegisterSink call, it would guarantee that cleanup is safer

Reply:
We could LOG_IF_FAILED here. I'm just worried about all these scope_exits not really making our code more simple.

Reply:
Added scope_exit after sink registrations